File: import/import.js
const readline = require('readline');
const fs = require('fs');

const terms = require('./terms');

terms.forEach(term => {

  let lines = [];

  let reader = readline.createInterface({
    input: fs.createReadStream(`./csv/${term.file}.csv`, {encoding: 'UTF-8'})
  });

  reader.on('line', line => {
    if (line != ',,,,,,,,,,' && !line.startsWith('CRN')) lines.push(line)
  });

  reader.on('close', function() {

    let courses = [];

    for (var i = 0; i < lines.length; i++) {

      let [crn, subject, number, name, credits, title,
        days, time, instructor, date, location] = lines[i].split(',');

      let meets = [{

        days: days,
        start_time: formatTime(time.split('-')[0]),
        end_time: formatTime(time.split('-')[1]),
        location: location,
        instructor: instructor.replace(';', ','),
        start_date: date.split('-')[0],
        end_date: date.split('-')[1]

      }];

      let tests = [];

      while (lines[i+1] && lines[i+1].startsWith(',')) {

        let [crn, subject, number, name, credits, title,
          days, time, instructor, date, location] = lines[++i].split(',');

        if (date.split('-')[0] == date.split('-')[1]) {

          tests.push({

            date: date.split('-')[0],
            start_time: formatTime(time.split('-')[0]),
            end_time: formatTime(time.split('-')[1]),
            location: location

          });

        } else {

          meets.push({

            days: days,
            start_time: formatTime(time.split('-')[0]),
            end_time: formatTime(time.split('-')[1]),
            location: location,
            instructor: instructor.replace(';', ','),
            start_date: date.split('-')[0],
            end_date: date.split('-')[1]

          });

        }

      }

      let course = courses.find(course => course.subject == subject && course.number == number);

      let section = {crn, name, tests, meets};

      if (course) {

        course.sections.push(section);

      } else {

        courses.push({semester_id: term.id, subject, number, title, credits, sections: [section]});

      }

    }

    let writer = fs.createWriteStream(`./json/${term.file}.json`);

    writer.write(JSON.stringify(courses));

  })

});

function formatTime(time) {

  if (!time) return;

  let parts = time.split(' ');

  let period = parts[1];
  let hours = parts[0].split(':')[0];
  let minutes = parts[0].split(':')[1];

  if (period == 'pm' && parseInt(hours) < 12) hours = parseInt(hours) + 12;

  return hours + ':' + minutes;

}

File: server/import.go
package main

import (
	"encoding/json"
	"io/ioutil"
)

type parsedCoursesJSON []struct {
	Subject  string
	Number   string
	Title    string
	Credits  string
	Sections []struct {
		CRN    string
		Name   string
		Campus string
		Tests  []struct {
			Date      string
			Location  string
			StartTime string
			EndTime   string
		}
		Meets []struct {
			Days       string
			StartTime  string
			EndTime    string
			Instructor string
			Location   string
			StartDate  string
			EndDate    string
		}
	}
}

func importCourses() {
	data, err := ioutil.ReadFile("import/json/spring-2017.json")
	if err != nil {
		handleError("Could not find sections.json")
	}

	var parsed parsedCoursesJSON
	json.Unmarshal(data, &parsed)
	//printInvalidCourses(&parsed)
	courses := getCoursesFromJSON(&parsed)
	sections := getSectionsFromJSON(&parsed)
	meets := getMeetsFromJSON(&parsed)
	batchInsertCourses(courses)
	batchInsertSections(sections)
	batchInsertMeets(meets)
}

func importTerms() {
	terms := []*Term{
		&Term{-1, "Fall", 2016, "Fall 2016"},
		&Term{-1, "Winter", 2016, "Winter 2016"},
		&Term{-1, "Spring", 2017, "Spring 2017"}}
	batchInsertTerms(terms)
}

//was used for parsing bad values, but its now done on front-end. Saving for future.
//TODO: move this to a utils.go file for future
/*
func printInvalidCourses(parsed *parsedCoursesJSON) {
	creditsRegex, _ := regexp.Compile("[\\d.-]")
	numberRegex, _ := regexp.Compile("\\d+([A-Z])?$")
	campusRegex, _ := regexp.Compile("^[LVOHM]$")
	sectionRegex, _ := regexp.Compile("[A-Z\\d]")
	subjectRegex, _ := regexp.Compile("[A-Z]{3}")
	titleRegex, _ := regexp.Compile("[A-Za-z\\s]+")
	instructorRegex, _ := regexp.Compile("[a-zA-Z()]")
	locationRegex, _ := regexp.Compile("[A-Za-z\\d+]")
	dateRegex, _ := regexp.Compile("\\d+\\/\\d+-\\d+\\/\\d+")
	daysRegex, _ := regexp.Compile("([MWTRFSU])+")
	timeRegex, _ := regexp.Compile("(\\d+:\\d+ \\s?(am|pm)-\\d+:\\d+ \\s?(am|pm))|TBA")

	for _, data := range *parsed {
		if !campusRegex.MatchString(data.Campus) {
			fmt.Printf("Invalid campus: %v\n", data.Campus)
		}
		if !creditsRegex.MatchString(data.Credits) {
			fmt.Printf("Invalid credits: %v\n", data.Credits)
		}
		if !numberRegex.MatchString(data.Number) {
			fmt.Printf("Invalid number: %v\n", data.Number)
		}
		if !sectionRegex.MatchString(data.Section) {
			fmt.Printf("Invalid section: %v\n", data.Section)
		}
		if !subjectRegex.MatchString(data.Subject) {
			fmt.Printf("Invalid subject: %v\n", data.Subject)
		}
		if !titleRegex.MatchString(data.Subject) {
			fmt.Printf("Invalid title: %v\n", data.Title)
		}
		for _, meet := range data.Meets {
			if !dateRegex.MatchString(meet.Date) {
				fmt.Printf("Invalid date: %v\n", meet.Date)
			}
			if meet.Days != "" && !daysRegex.MatchString(meet.Days) {
				fmt.Printf("Invalid days: %v\n", meet.Days)
			}
			if !instructorRegex.MatchString(meet.Instructor) {
				fmt.Printf("Invalid instructor: %v\n", meet.Instructor)
			}
			if !locationRegex.MatchString(meet.Location) {
				fmt.Printf("Invalid location: %v\n", meet.Location)
			}
			if meet.Time != "" && !timeRegex.MatchString(meet.Time) {
				fmt.Printf("Invalid time: %v\n", meet.Time)
			}
		}
	}
}*/
